modules/standard/GPU.chpl

```diff
@@ -626,4 +626,191 @@ module GPU
   */
   inline proc gpuMaxLocReduce(const ref A: [] ?t) do return doGpuReduce("maxloc", A);
 
+
+
+  // ============================
+  // GPU Scans
+  // ============================
+
+  // The following functions are used to implement GPU scans. They are
+  // intended to be called from a GPU locale.
+  import BitOps;
+  import Math;
+
+
+  proc gpuScan(ref gpuArr: [] uint){
+    if gpuArr.size==0 then return;
+    // Use a simple algorithm for small arrays
+    // TODO check the actual thread block size rather than the2*default
+    if gpuArr.size <= 1024 {
+      // The algorithms only works for arrays that are size of a power of two.
+      // In case it's not a power of two we pad it out with 0s
+      const size = roundToPowerof2(gpuArr.size);
+      var arr : [0..<size] uint;
+      arr[0..<gpuArr.size] = gpuArr;
+
+      // Hillis Steele Scan is better if we can scan in
+      // a single thread block
+      // TODO check the actual thread block size rather than the default
+      if gpuArr.size <= 512 then
+        hillisSteeleScan(arr);
+      else
+        blellochScan(arr);
+
+      // Copy back
```

Comment:
Can we avoid copying in cases where `gpuArr.size` is already a power of two (and thus we don't need to pad it)?

This function gets called recursively by `parallelArrScan` right, so it's probably worth trying to optimize.